File: ios/Plugin/InlinePDFPlugin.swift
import Foundation
import Capacitor
import PDFKit
import WebKit

@objc(InlinePDFPlugin)
public class InlinePDFPlugin: CAPPlugin {
    private var pdfViews: [String: InlinePDFViewController] = [:]
    
    @objc func create(_ call: CAPPluginCall) {
        print("InlinePDF: create called")
            
        guard let containerId = call.getString("containerId") else {
            print("InlinePDF: Missing containerId")
            call.reject("Missing containerId")
            return
        }
        
        guard let rect = call.getObject("rect") else {
            print("InlinePDF: Missing rect")
            call.reject("Missing rect")
            return
        }
        
        print("InlinePDF: rect = \(rect)")
        
        let x = rect["x"] as? Double ?? 0
        let y = rect["y"] as? Double ?? 0
        let width = rect["width"] as? Double ?? 0
        let height = rect["height"] as? Double ?? 0
        
        print("InlinePDF: Creating PDF view with rect: x=\(x), y=\(y), width=\(width), height=\(height)")
        
        let viewerId = UUID().uuidString
        
        // Must create UI components on main thread
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                call.reject("Plugin instance not available")
                return
            }
            
            let pdfVC = InlinePDFViewController()
            
            // Configure initial settings
            if let bgColor = call.getString("backgroundColor") {
                pdfVC.backgroundColor = bgColor
            }
            
            if let initialScale = call.getFloat("initialScale") {
                pdfVC.initialScale = CGFloat(initialScale)
            }
            
            // Get the view controller container
            guard let bridge = self.bridge,
                  let webView = bridge.webView,
                  let viewController = bridge.viewController else {
                print("InlinePDF: Could not access bridge components")
                call.reject("Could not access bridge components")
                return
            }
            
            print("InlinePDF: Got bridge components")
            
            // Properly add as child view controller
            viewController.addChild(pdfVC)
            pdfVC.view.backgroundColor = UIColor.white
            pdfVC.view.frame = CGRect(x: CGFloat(x), y: CGFloat(y), width: CGFloat(width), height: CGFloat(height))
            
            print("InlinePDF: Adding PDF view to web view")
            webView.addSubview(pdfVC.view)
            pdfVC.didMove(toParent: viewController)
            print("InlinePDF: Added PDF view to web view")
            
            // Set up callbacks
            pdfVC.onGestureStart = { [weak self] in
                self?.notifyListeners("gestureStart", data: [:])
            }
            
            pdfVC.onGestureEnd = { [weak self] in
                self?.notifyListeners("gestureEnd", data: [:])
            }
            
            pdfVC.onPageChanged = { [weak self] page in
                self?.notifyListeners("pageChanged", data: ["page": page])
            }
            
            pdfVC.onZoomChanged = { [weak self] zoom in
                self?.notifyListeners("zoomChanged", data: ["zoom": zoom])
            }
            
            self.pdfViews[viewerId] = pdfVC
            
            print("InlinePDF: Stored PDF view with id: \(viewerId)")
            print("InlinePDF: Resolving call")
            
            call.resolve(["viewerId": viewerId])
            
            print("InlinePDF: create completed")
        }
    }
    
    @objc func loadPDF(_ call: CAPPluginCall) {
        print("InlinePDF: loadPDF called")
        
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId] else {
            print("InlinePDF: Invalid viewer ID")
            call.reject("Invalid viewer ID")
            return
        }
        
        print("InlinePDF: Found PDF view for id: \(viewerId)")
        
        if let urlString = call.getString("url") {
            // Handle Capacitor file URLs
            let fileUrl: URL?
            if urlString.hasPrefix("capacitor://localhost/_capacitor_file_") {
                // Extract the actual file path from Capacitor URL
                let path = urlString
                    .replacingOccurrences(of: "capacitor://localhost/_capacitor_file_", with: "")
                    .removingPercentEncoding ?? ""
                fileUrl = URL(fileURLWithPath: path)
            } else if let url = URL(string: urlString) {
                fileUrl = url
            } else {
                fileUrl = nil
            }
            
            if let url = fileUrl {
                print("InlinePDF: Loading PDF from URL: \(url)")
                pdfVC.loadPDF(from: url)
                call.resolve()
                print("InlinePDF: loadPDF completed")
            } else {
                print("InlinePDF: Invalid URL format")
                call.reject("Invalid URL format")
            }
        } else if let path = call.getString("path") {
            let url = URL(fileURLWithPath: path)
            pdfVC.loadPDF(from: url)
            call.resolve()
        } else {
            call.reject("No URL or path provided")
        }
    }
    
    @objc func search(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId],
              let query = call.getString("query") else {
            call.reject("Invalid parameters")
            return
        }
        
        let caseSensitive = call.getBool("caseSensitive") ?? false
        let wholeWords = call.getBool("wholeWords") ?? false
        
        DispatchQueue.main.async {
            let results = pdfVC.search(query, caseSensitive: caseSensitive, wholeWords: wholeWords)
            let resultData = results.map { result in
                return [
                    "page": result.page,
                    "text": result.text,
                    "bounds": [
                        "x": result.bounds.origin.x,
                        "y": result.bounds.origin.y,
                        "width": result.bounds.width,
                        "height": result.bounds.height
                    ],
                    "context": result.context ?? ""
                ]
            }
            
            call.resolve(["results": resultData])
        }
    }
    
    @objc func goToPage(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId],
              let page = call.getInt("page") else {
            call.reject("Invalid parameters")
            return
        }
        
        let animated = call.getBool("animated") ?? true
        
        DispatchQueue.main.async {
            pdfVC.goToPage(page, animated: animated)
            call.resolve()
        }
    }
    
    @objc func getState(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId] else {
            call.reject("Invalid viewer ID")
            return
        }
        
        DispatchQueue.main.async {
            let state = pdfVC.getState()
            call.resolve([
                "currentPage": state.currentPage,
                "totalPages": state.totalPages,
                "zoom": state.zoom,
                "isLoading": state.isLoading
            ])
        }
    }
    
    @objc func updateRect(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId],
              let rect = call.getObject("rect"),
              let x = rect["x"] as? CGFloat,
              let y = rect["y"] as? CGFloat,
              let width = rect["width"] as? CGFloat,
              let height = rect["height"] as? CGFloat else {
            call.reject("Invalid parameters")
            return
        }
        
        DispatchQueue.main.async {
            pdfVC.updateRect(CGRect(x: x, y: y, width: width, height: height))
            call.resolve()
        }
    }
    
    @objc func resetZoom(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId] else {
            call.reject("Invalid viewer ID")
            return
        }
        
        DispatchQueue.main.async {
            pdfVC.resetZoom()
            call.resolve()
        }
    }
    
    @objc func clearHighlights(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId] else {
            call.reject("Invalid viewer ID")
            return
        }
        
        DispatchQueue.main.async {
            pdfVC.clearSearchHighlights()
            call.resolve()
        }
    }
    
    @objc func destroy(_ call: CAPPluginCall) {
        guard let viewerId = call.getString("viewerId"),
              let pdfVC = pdfViews[viewerId] else {
            call.reject("Invalid viewer ID")
            return
        }
        
        DispatchQueue.main.async {
            print("InlinePDF: Destroying PDF view with id: \(viewerId)")
            
            // Properly remove child view controller
            pdfVC.willMove(toParent: nil)
            pdfVC.view.removeFromSuperview()
            pdfVC.removeFromParent()
            
            // Force cleanup of the PDF document
            if let pdfView = pdfVC.view.subviews.first(where: { $0 is PDFView }) as? PDFView {
                pdfView.document = nil
                pdfView.removeFromSuperview()
            }
            
            self.pdfViews.removeValue(forKey: viewerId)
            print("InlinePDF: PDF view destroyed successfully")
            call.resolve()
        }
    }
}

File: android/src/main/java/com/protean/capacitor/inlinepdf/InlinePDFPlugin.kt
package com.protean.capacitor.inlinepdf

import android.graphics.Color
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.UUID

@CapacitorPlugin(name = "InlinePDF")
class InlinePDFPlugin : Plugin() {
    private val pdfViews = mutableMapOf<String, InlinePDFView>()
    
    @PluginMethod
    fun create(call: PluginCall) {
        val containerId = call.getString("containerId") ?: run {
            call.reject("Missing containerId")
            return
        }
        
        val rect = call.getObject("rect") ?: run {
            call.reject("Missing rect")
            return
        }
        
        val x = rect.getDouble("x")?.toFloat() ?: 0f
        val y = rect.getDouble("y")?.toFloat() ?: 0f
        val width = rect.getDouble("width")?.toInt() ?: 0
        val height = rect.getDouble("height")?.toInt() ?: 0
        
        val viewerId = UUID.randomUUID().toString()
        
        activity.runOnUiThread {
            try {
                // Create PDF view
                val pdfView = InlinePDFView(context).apply {
                    layoutParams = CoordinatorLayout.LayoutParams(width, height).apply {
                        leftMargin = x.toInt()
                        topMargin = y.toInt()
                    }

                    // CRITICAL: Enable touch event handling
                    isClickable = true
                    isFocusable = true
                    isFocusableInTouchMode = true

                    // Set background color if specified
                    call.getString("backgroundColor")?.let { bgColor ->
                        try {
                            setBackgroundColor(Color.parseColor(bgColor))
                        } catch (e: Exception) {
                            // Ignore invalid color
                        }
                    }

                    // Set initial scale if specified
                    call.getFloat("initialScale")?.let { scale ->
                        initialScale = scale
                    }
                }
                
                // Set up callbacks
                pdfView.onGestureStart = {
                    notifyListeners("gestureStart", JSObject())
                }
                
                pdfView.onGestureEnd = {
                    notifyListeners("gestureEnd", JSObject())
                }
                
                pdfView.onPageChanged = { page ->
                    notifyListeners("pageChanged", JSObject().apply {
                        put("page", page)
                    })
                }
                
                pdfView.onZoomChanged = { zoom ->
                    notifyListeners("zoomChanged", JSObject().apply {
                        put("zoom", zoom)
                    })
                }
                
                // Add to web view container
                val webView = bridge.webView
                val parent = webView.parent as? ViewGroup

                if (parent != null) {
                    parent.addView(pdfView)
                    // Bring PDF view to front to ensure it receives touch events
                    pdfView.bringToFront()
                } else {
                    // If parent is null, add directly to activity's content view
                    activity.addContentView(pdfView, pdfView.layoutParams)
                    pdfView.bringToFront()
                }

                pdfViews[viewerId] = pdfView
                
                call.resolve(JSObject().apply {
                    put("viewerId", viewerId)
                })
            } catch (e: Exception) {
                call.reject("Failed to create PDF view", e)
            }
        }
    }
    
    @PluginMethod
    fun loadPDF(call: PluginCall) {
        val viewerId = call.getString("viewerId") ?: run {
            call.reject("Missing viewerId")
            return
        }
        
        val pdfView = pdfViews[viewerId] ?: run {
            call.reject("Invalid viewer ID")
            return
        }
        
        activity.runOnUiThread {
            try {
                when {
                    call.getString("url") != null -> {
                        val url = call.getString("url")!!
                        pdfView.loadFromUrl(url)
                        call.resolve()
                    }
                    call.getString("path") != null -> {
                        val path = call.getString("path")!!
                        pdfView.loadFromPath(path)
                        call.resolve()
                    }
                    else -> {
                        call.reject("No URL or path provided")
                    }
                }
            } catch (e: Exception) {
                call.reject("Failed to load PDF", e)
            }
        }
    }
    
    @PluginMethod
    fun search(call: PluginCall) {
        val viewerId = call.getString("viewerId") ?: run {
            call.reject("Missing viewerId")
            return
        }
        
        val query = call.getString("query") ?: run {
            call.reject("Missing query")
            return
        }
        
        val pdfView = pdfViews[viewerId] ?: run {
            call.reject("Invalid viewer ID")
            return
        }
        
        val caseSensitive = call.getBoolean("caseSensitive", false) ?: false
        val wholeWords = call.getBoolean("wholeWords", false) ?: false
        
        activity.runOnUiThread {
            try {
                val results = pdfView.search(query, caseSensitive, wholeWords)
                val resultsArray = JSArray()
                
                results.forEach { result ->
                    resultsArray.put(JSObject().apply {
                        put("page", result.page)
                        put("text", result.text)
                        put("bounds", JSObject().apply {
                            put("x", result.bounds.left)
                            put("y", result.bounds.top)
                            put("width", result.bounds.width())
                            put("height", result.bounds.height())
                        })
                        result.context?.let { put("context", it) }
                    })
                }
                
                call.resolve(JSObject().apply {
                    put("results", resultsArray)
                })
            } catch (e: Exception) {
                call.reject("Search failed", e)
            }
        }
    }
    
    @PluginMethod
    fun goToPage(call: PluginCall) {
        val viewerId = call.getString("viewerId") ?: run {
            call.reject("Missing viewerId")
            return
        }
        
        val page = call.getInt("page") ?: run {
            call.reject("Missing page")
            return
        }
        
        val pdfView = pdfViews[viewerId] ?: run {
            call.reject("Invalid viewer ID")
            return
        }
        
        val animated = call.getBoolean("animated", true) ?: true
        
        activity.runOnUiThread {
            try {
                pdfView.goToPage(page, animated)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to navigate to page", e)
            }
        }
    }
    
    @PluginMethod
    fun getState(call: PluginCall) {
        val viewerId = call.getString("viewerId") ?: run {
            call.reject("Missing viewerId")
            return
        }
        
        val pdfView = pdfViews[viewerId] ?: run {
            call.reject("Invalid viewer ID")
            return
        }
        
        activity.runOnUiThread {
            try {
                val state = pdfView.getState()
                call.resolve(JSObject().apply {
                    put("currentPage", state.currentPage)
                    put("totalPages", state.totalPages)
                    put("zoom", state.zoom)
                    put("isLoading", state.isLoading)
                })
            } catch (e: Exception) {
                call.reject("Failed to get state", e)
            }
        }
    }
    
    @PluginMethod
    fun updateRect(call: PluginCall) {
        val viewerId = call.getString("viewerId") ?: run {
            call.reject("Missing viewerId")
            return
        }
        
        val rect = call.getObject("rect") ?: run {
            call.reject("Missing rect")
            return
        }
        
        val pdfView = pdfViews[viewerId] ?: run {
            call.reject("Invalid viewer ID")
            return
        }
        
        val x = rect.getDouble("x")?.toFloat() ?: 0f
        val y = rect.getDouble("y")?.toFloat() ?: 0f
        val width = rect.getDouble("width")?.toInt() ?: 0
        val height = rect.getDouble("height")?.toInt() ?: 0
        
        activity.runOnUiThread {
            try {
                pdfView.layoutParams = CoordinatorLayout.LayoutParams(width, height).apply {
                    leftMargin = x.toInt()
                    topMargin = y.toInt()
                }
                pdfView.requestLayout()
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to update rect", e)
            }
        }
    }
    
    @PluginMethod
    fun destroy(call: PluginCall) {
        val viewerId = call.getString("viewerId") ?: run {
            call.reject("Missing viewerId")
            return
        }
        
        val pdfView = pdfViews[viewerId] ?: run {
            call.reject("Invalid viewer ID")
            return
        }
        
        activity.runOnUiThread {
            try {
                (pdfView.parent as? ViewGroup)?.removeView(pdfView)
                pdfView.cleanup()
                pdfViews.remove(viewerId)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to destroy viewer", e)
            }
        }
    }
    
    override fun handleOnDestroy() {
        super.handleOnDestroy()
        // Clean up all PDF views
        pdfViews.values.forEach { pdfView ->
            (pdfView.parent as? ViewGroup)?.removeView(pdfView)
            pdfView.cleanup()
        }
        pdfViews.clear()
    }
}